### PR TITLE
feat[CL]: lock migration to frozen concentrated position

### PR DESCRIPTION
**Pull request by czarcas7ic**
## What is the purpose of the change

We currently only have logic to migrate unlocked LP tokens to a full range CL position. This PR introduces logic slightly similar to the unpool logic, where we 

1. Superfluid undelegate if lock was superfluid delegated
2. Burn underlying synthetic OSMO
3. Break the lock

But now instead of creating new locks, we simply create full range CL position that is frozen for either the entire lock duration, or the remaining lock duration if it was in the process of unlocking.

## Brief Changelog

- Adds clKeeper dependence to the superfluid keeper
- Removes pool_id_entering proto field (is not automatically discerned from gov sanctioned link)
- Adds `MsgUnlockAndMigrateSharesToFullRangeConcentratedPosition`
  - This is used for locked LP tokens, whereas `MsgMigrateSharesToFullRangeConcentratedPosition` in gamm is only used on unlocked LP tokens
- Adds test cases for new `UnlockAndMigrate` logic 
- Adds event for `UnlockAndMigrate`

## Testing and Verifying

Various test cases added as described above.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: proto/osmosis/gamm/pool-models/balancer/tx/tx.proto
Comment (czarcas7ic) [inline proto/osmosis/gamm/pool-models/balancer/tx/tx.proto:46]: Note to reviewer: No longer required since we now get this from the gov sanctioned link between pools.

File: x/gamm/keeper/migrate.go
Comment (czarcas7ic) [inline x/gamm/keeper/migrate.go:49]: Note to reviewer: We freeze for the zero value of time here since only unlocked LP tokens can get migrated through this entry point.

File: x/gamm/keeper/migrate_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/migrate_test.go:119]: Note to reviewer: This test case is no longer possible due to user no longer being the one to provide link information

File: x/gamm/keeper/pool.go
Comment (czarcas7ic) [inline x/gamm/keeper/pool.go:308]: Note to reviewer: Removed since this is code duplication. Used the original func instead.

File: x/superfluid/keeper/migrate_test.go
Comment (mattverse) [inline x/superfluid/keeper/migrate_test.go:185]: Do you have a brief number of how much the joinPoolAmt is here? (Asking to get a brief sense of the significance of the tolerance here)

File: x/superfluid/keeper/migrate_test.go
Comment (mattverse) [inline x/superfluid/keeper/migrate_test.go:190]: Can we add a check that checks if the frozen unitl has been properly migrated?

File: x/superfluid/types/expected_keepers.go
Comment (mattverse) [inline x/superfluid/types/expected_keepers.go:52]: Why do we need gamm keeper's Migrate method?

**Review by mattverse — COMMENTED** [review]
Nice work! Thanks for all the detailed in line comments, was super helpful in reviewing this PR. Left some minor comments, please take a look.

File: x/superfluid/types/expected_keepers.go
Comment (czarcas7ic) [reply x/superfluid/types/expected_keepers.go:52]: We don't, good catch! Before I wanted to make superfluid the only entrypoint and simply skip the unlocking part if it wasn't locked, but it ended up being a bad design and I scrapped it but left this behind. Removed.

File: x/gamm/keeper/migrate.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate.go]: Changed name as requested here https://github.com/osmosis-labs/osmosis/pull/4129/commits/36fb46c060e7274b5026b87b2d3632b38c9dffa4

File: x/gamm/keeper/migrate.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate.go:20]: Do you think I should reintroduce the function I removed from this PR? I basically wanted to utilize a function that already existed but I realize here that we panic if the denom does not exist. Is this problematic? If so, I will reintroduce the function I had made previously for this.

File: x/gamm/keeper/migrate_test.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate_test.go]: Added check here https://github.com/osmosis-labs/osmosis/pull/4129/commits/e9412d780285a5fc6dab175e080ca0911f9aa739

File: x/gamm/keeper/migrate.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate.go:204]: That is done in [this PR](https://github.com/osmosis-labs/osmosis/pull/4126/files) and I intend to merge it before this one. Good catch though

File: x/superfluid/keeper/migrate.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: https://github.com/osmosis-labs/osmosis/pull/4129/commits/a06774daf043b4f8e8803dc7ebb34a8be92761f6

File: x/gamm/keeper/migrate_test.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate_test.go]: And here in superfluid https://github.com/osmosis-labs/osmosis/pull/4129/commits/c4165de16b24c7c541bb83e5baa4257ce1c1b549

File: x/superfluid/keeper/migrate_test.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate_test.go:203]: What I ended up doing was checking that the expected position with a frozenUntil value was created https://github.com/osmosis-labs/osmosis/pull/4129/commits/5f0d2a048e2e2df35e8866eabf91cbcbf24210e6

We test `getExistingLockRemainingDuration` on its own, so we have a guarantee that whatever that frozenUntil value that gets returned is valid. Your point still stands though and it is now indirectly checked!

File: x/superfluid/keeper/migrate_test.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate_test.go:198]: Yeah, these values are equal to 5000 uosmo so the tolerance would be 2 uosmo compared to that. To be clear though (and this originally wasn't clear to me before) the user doesn't get shorted these 2 uosmo, they simply remain in the wallet and are free from any lock. Even with very large amounts this amounts to dust though.

File: x/gamm/keeper/migrate.go
Comment (p0mvn) [inline x/gamm/keeper/migrate.go:199]: Spec doesn't apply to method

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate.go:15]:
```suggestion
// If the lock is superfluid delegated, it will undelegate the superfluid position.
```

Comment (p0mvn) [inline x/superfluid/keeper/migrate.go:15]: Consider elaborating when errors are expected

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate.go:35]: We might want to have a helper in CL module that returns the desired type. This seems to be common enough to justify this helper and replace of all of the type casts

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate.go:39]: How do we compare the correct LP shares part? That is, how do we know that the lock corresponds to the right amount of shares?

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate.go:38]: What does it mean? I looked at the caller and don't see the same validation being made anywhere

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [reply x/superfluid/keeper/migrate.go:39]: I think we need a check and a test that `sharesToMigrate` and lock's shares are the same

File: x/superfluid/keeper/migrate_test.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate_test.go:133]: Let's add a test where lock has a different number of shares locked from what we are trying to migrate

File: x/gamm/keeper/migrate.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate.go]: https://github.com/osmosis-labs/osmosis/pull/4129/commits/4f62673f13b377e76a5feecccd511c09e4c08f9c

File: x/superfluid/types/expected_keepers.go
Comment (czarcas7ic) [reply x/superfluid/types/expected_keepers.go]: Renamed https://github.com/osmosis-labs/osmosis/pull/4129/commits/786389c534947066e7fadd7e30a99005825ad662

File: x/gamm/keeper/migrate.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate.go]: Added https://github.com/osmosis-labs/osmosis/pull/4129/commits/46404a807264638c2071b92d4840714e739023a0

File: x/superfluid/keeper/migrate.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: Added https://github.com/osmosis-labs/osmosis/pull/4129/commits/8bdae4d6612b94284f3669241dd7814a5fae8945

File: x/superfluid/keeper/migrate.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: Added helper function in CL and implemented in gamm and superfluid https://github.com/osmosis-labs/osmosis/pull/4129/commits/7daae195de526ec1dcffc273516bccc83be9c560

File: x/superfluid/keeper/migrate.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: You are correct, it is not expected to be true by the caller. I utilized `validateLockForUnpool` from the unpool logic to prevent code duplication but the comment should have been changed. I changed the comment here, thanks! https://github.com/osmosis-labs/osmosis/pull/4129/commits/738b4a994d95e8cfc5c740d3fd965b2e6188505f

File: x/superfluid/keeper/migrate_test.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate_test.go]: Great point, I modified all the tests here to use varying percentages of the initial join amount! https://github.com/osmosis-labs/osmosis/pull/4129/commits/d961640126dbf52f7b779ae531a710723ecb92e3

File: x/superfluid/keeper/migrate.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: CL shares and balancer shares are not 1:1. With that being said, we test what the shares represent in terms of the underlying which is all that really matters IMO. Let me know if this addresses that concern.

Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: Ohh I think I see the confusion now, the comment says "contains correct LP shares" but what it actually means is "contains correct denom LP shares".

Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: Added this clarifying comment, lmk if that is sufficient for you: https://github.com/osmosis-labs/osmosis/pull/4129/commits/28f04163b4c7d1bccd2021d44dc209d4f79b9af6

File: x/gamm/keeper/migrate.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate.go]: Agreed, did this with `CreateFullRangePosition` and `GetPoolFromPoolIdAndConvertToConcentrated`. While I could have actually put the `GetPoolFromPoolIdAndConvertToConcentrated` logic into `CreateFullRangePosition`, I want to have the poolId check in the very beginning to prevent overhead when running in invalid tx.

https://github.com/osmosis-labs/osmosis/pull/4129/commits/788065d31ce523e812cc679013c9dbde29ea6458

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [reply x/superfluid/keeper/migrate.go]: I guess my main question is with regards to the number of shares locked being greater than the number of shares we migrate.

Say we migrate 4 shares but the lock contains 10. Then, we exit 4/10 shares. What happens to the other 6? Currently, it seems that they are left unlocked.

Maybe we should relock them?

Also, could you please document in the method spec and test the final behavior of this excess lock share amount

File: x/gamm/keeper/migrate.go
Comment (mattverse) [reply x/gamm/keeper/migrate.go]: Sweet thanks!

File: x/gamm/keeper/migrate.go
Comment (mattverse) [reply x/gamm/keeper/migrate.go:20]: Maybe it's a open question, but I personally see that unifying error / panic as much as possible, and letting the caller decide what to do with the error as a better design, so that we don't have multiple places that errors or panics.(Setters and getters from state are unevitable for panicing). But again, its an open ended question, I'll leave it up to you to decide on this :)

Comment (mattverse) [reply x/gamm/keeper/migrate.go:20]: I personally think unifying panics and errors as much as possible is a better design so that we don't have multiple places here and there that either errors or panics, so that the caller of this function can have a unified point of handling errors (the must getters and must setters from state is inevitable) . But again, I guess this is an open ended question, I'll leave it to you to decide :)

File: x/superfluid/keeper/migrate_test.go
Comment (mattverse) [reply x/superfluid/keeper/migrate_test.go:206]: Sweet! Thanks for clarifying

**Review by mattverse — APPROVED** [review]
my main concerns and reviews have all been addressed! LGTM 🌮

File: x/gamm/keeper/migrate.go
Comment (czarcas7ic) [reply x/gamm/keeper/migrate.go]: Agreed! Addressed here https://github.com/osmosis-labs/osmosis/pull/4129/commits/d3bdaa11aa1768044f843bd148388a498601c318

File: x/superfluid/keeper/migrate.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: Woh, this is a great point. Thanks for pointing this out! Let me address it before we merge this

Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go]: @p0mvn I believe the following commit solves the issue you found, thanks for pointing this out!

https://github.com/osmosis-labs/osmosis/pull/4129/commits/68c7991099c48746ef1fb25c640cb7b04e3328af

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate.go:63]: nit: can we move this further down to line 86 where this is being used for the first time

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate.go:95]: nit: we can call `beginUnlock` directly

File: x/superfluid/keeper/migrate.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate.go:93]: How is beginning to unlock where it left off guaranteed?

File: x/superfluid/keeper/migrate_test.go
Comment (p0mvn) [inline x/superfluid/keeper/migrate_test.go:62]: Is it possible to test the "where it left" off part somehow>?

**Review by p0mvn — APPROVED** [review]
LGTM if my questions have been considered

File: x/superfluid/keeper/migrate.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go:95]: I don't think we can, unless we want to export beginUnlock from the lockup module.

File: x/superfluid/keeper/migrate.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go:93]: We lock it above with freezeDuration which is calculated as the remaining time the lock had before breaking, but I agree this needs to be tested

Comment (czarcas7ic) [reply x/superfluid/keeper/migrate.go:93]: Added test! https://github.com/osmosis-labs/osmosis/pull/4129/commits/04518037f42656afd74f111b82f9bdbfe96b541b

File: x/superfluid/keeper/migrate_test.go
Comment (czarcas7ic) [reply x/superfluid/keeper/migrate_test.go:62]: Added test! https://github.com/osmosis-labs/osmosis/pull/4129/commits/04518037f42656afd74f111b82f9bdbfe96b541b

**Pull request merged**
